File: 0967-minimum-falling-path-sum/0967-minimum-falling-path-sum.cpp
class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& matrix) {
        int n = matrix.size();
        int m = matrix[0].size();

        vector<vector<int>> dp(n, vector<int>(m, 0));

        // First row stays as is
        for (int j = 0; j < m; j++) {
            dp[0][j] = matrix[0][j];
        }

        for (int i = 1; i < n; i++) {
            for (int j = 0; j < m; j++) {
                int up = dp[i - 1][j];
                int leftDiag = j > 0 ? dp[i - 1][j - 1] : 1e9;
                int rightDiag = j < m - 1 ? dp[i - 1][j + 1] : 1e9;

                dp[i][j] = matrix[i][j] + min({up, leftDiag, rightDiag});
            }
        }

        // Take min of last row
        int ans = INT_MAX;
        for (int j = 0; j < m; j++) {
            ans = min(ans, dp[n - 1][j]);
        }

        return ans;
    }
};

File: 0967-minimum-falling-path-sum/0967-minimum-falling-path-sum.java
class Solution {
    int[][] matrix;
    int n;
    Integer[][] memo;               // null => uncomputed
    static final int INF = 1_000_000_000;

    int solve(int r, int c){
        if (c < 0 || c >= n || r >= n) return INF;   // invalid path
        if (r == n - 1) return matrix[r][c];

        if (memo[r][c] != null) return memo[r][c];

        int down = solve(r + 1, c);
        int dr   = solve(r + 1, c + 1);
        int dl   = solve(r + 1, c - 1);

        int best = Math.min(down, Math.min(dr, dl));
        return memo[r][c] = matrix[r][c] + best;     // best is large but safe (no overflow)
    }

    public int minFallingPathSum(int[][] matrix) {
        this.matrix = matrix;
        this.n = matrix.length;
        this.memo = new Integer[n][n];

        int ans = Integer.MAX_VALUE;
        for (int c = 0; c < n; c++) {
            ans = Math.min(ans, solve(0, c));
        }
        return ans;
    }
}
